{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'C:/Users/heung/OneDrive/Documents/scData/cgga_mRNAseq_325/CGGA.mRNAseq_325_clinical.20200506.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Load the dataset\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m df_clinical \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mC:/Users/heung/OneDrive/Documents/scData/cgga_mRNAseq_325/CGGA.mRNAseq_325_clinical.20200506.txt\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m df_clinical\u001b[38;5;241m.\u001b[39mrename(columns \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCensor (alive=0; dead=1)\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCensor\u001b[39m\u001b[38;5;124m'\u001b[39m}, inplace \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      5\u001b[0m columns_to_keep \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCGGA_ID\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPRS_type\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGrade\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mOS\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCensor\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[0;32m~/Desktop/bioinformatics/jswenv/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1013\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m   1014\u001b[0m     dialect,\n\u001b[1;32m   1015\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1022\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m   1023\u001b[0m )\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/bioinformatics/jswenv/lib/python3.12/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    619\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    623\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/Desktop/bioinformatics/jswenv/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/bioinformatics/jswenv/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1880\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1878\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1879\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1880\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1882\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1883\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1884\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1885\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1886\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1887\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1888\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1889\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1890\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1891\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/Desktop/bioinformatics/jswenv/lib/python3.12/site-packages/pandas/io/common.py:873\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    868\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    869\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    870\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    871\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    872\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 873\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    874\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    875\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    876\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    877\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    878\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    879\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    880\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    881\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    882\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'C:/Users/heung/OneDrive/Documents/scData/cgga_mRNAseq_325/CGGA.mRNAseq_325_clinical.20200506.txt'"
     ]
    }
   ],
   "source": [
    "\n",
    "# Load the dataset\n",
    "df_clinical = pd.read_csv(\"C:/Users/heung/OneDrive/Documents/scData/cgga_mRNAseq_325/CGGA.mRNAseq_325_clinical.20200506.txt\", sep=\"\\t\")\n",
    "df_clinical.rename(columns = {'Censor (alive=0; dead=1)': 'Censor'}, inplace = True)\n",
    "\n",
    "columns_to_keep = ['CGGA_ID', 'PRS_type', 'Grade', 'OS', 'Censor']\n",
    "df_clinical = df_clinical[columns_to_keep]\n",
    "df_clinical.tail(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PRS_type\n",
       "Primary      229\n",
       "Recurrent     62\n",
       "Secondary     30\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinical['PRS_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_temp = pd.read_excel('gain_geneset.xlsx')\n",
    "\n",
    "gain_genes = df_temp.iloc[:, 0].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1154"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gain_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGGA_ID</th>\n",
       "      <th>AASS</th>\n",
       "      <th>ABCA13</th>\n",
       "      <th>ABCB1</th>\n",
       "      <th>ABCB4</th>\n",
       "      <th>ABCB5</th>\n",
       "      <th>ABCB8</th>\n",
       "      <th>ABCF2</th>\n",
       "      <th>ABHD11</th>\n",
       "      <th>ACHE</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF862</th>\n",
       "      <th>ZNF890P</th>\n",
       "      <th>ZNF92</th>\n",
       "      <th>ZNHIT1</th>\n",
       "      <th>ZNRF2</th>\n",
       "      <th>ZNRF2P1</th>\n",
       "      <th>ZNRF2P2</th>\n",
       "      <th>ZP3</th>\n",
       "      <th>ZSCAN21</th>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>CGGA_243</td>\n",
       "      <td>7.39</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8.22</td>\n",
       "      <td>1.55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.81</td>\n",
       "      <td>24.37</td>\n",
       "      <td>10.29</td>\n",
       "      <td>3.39</td>\n",
       "      <td>...</td>\n",
       "      <td>7.75</td>\n",
       "      <td>0.11</td>\n",
       "      <td>14.08</td>\n",
       "      <td>24.01</td>\n",
       "      <td>3.48</td>\n",
       "      <td>0.32</td>\n",
       "      <td>2.47</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.53</td>\n",
       "      <td>21.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>CGGA_247</td>\n",
       "      <td>18.09</td>\n",
       "      <td>0.29</td>\n",
       "      <td>6.13</td>\n",
       "      <td>1.03</td>\n",
       "      <td>0.09</td>\n",
       "      <td>13.36</td>\n",
       "      <td>22.01</td>\n",
       "      <td>11.25</td>\n",
       "      <td>2.66</td>\n",
       "      <td>...</td>\n",
       "      <td>48.01</td>\n",
       "      <td>0.85</td>\n",
       "      <td>14.51</td>\n",
       "      <td>29.08</td>\n",
       "      <td>3.64</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.74</td>\n",
       "      <td>9.18</td>\n",
       "      <td>67.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>CGGA_738</td>\n",
       "      <td>22.54</td>\n",
       "      <td>1.01</td>\n",
       "      <td>15.89</td>\n",
       "      <td>1.53</td>\n",
       "      <td>0.19</td>\n",
       "      <td>27.94</td>\n",
       "      <td>28.72</td>\n",
       "      <td>17.74</td>\n",
       "      <td>18.81</td>\n",
       "      <td>...</td>\n",
       "      <td>15.13</td>\n",
       "      <td>0.32</td>\n",
       "      <td>6.47</td>\n",
       "      <td>55.57</td>\n",
       "      <td>5.67</td>\n",
       "      <td>0.69</td>\n",
       "      <td>3.64</td>\n",
       "      <td>0.45</td>\n",
       "      <td>10.07</td>\n",
       "      <td>72.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>CGGA_759</td>\n",
       "      <td>16.61</td>\n",
       "      <td>1.08</td>\n",
       "      <td>14.85</td>\n",
       "      <td>1.14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.81</td>\n",
       "      <td>21.38</td>\n",
       "      <td>14.9</td>\n",
       "      <td>13.82</td>\n",
       "      <td>...</td>\n",
       "      <td>12.43</td>\n",
       "      <td>0.25</td>\n",
       "      <td>11.28</td>\n",
       "      <td>84.91</td>\n",
       "      <td>7.98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.13</td>\n",
       "      <td>0.43</td>\n",
       "      <td>10.12</td>\n",
       "      <td>86.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>CGGA_D30</td>\n",
       "      <td>5.57</td>\n",
       "      <td>0.15</td>\n",
       "      <td>14.73</td>\n",
       "      <td>1.89</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.32</td>\n",
       "      <td>21.45</td>\n",
       "      <td>18.22</td>\n",
       "      <td>12.18</td>\n",
       "      <td>...</td>\n",
       "      <td>7.77</td>\n",
       "      <td>0.27</td>\n",
       "      <td>5.89</td>\n",
       "      <td>114.33</td>\n",
       "      <td>8.57</td>\n",
       "      <td>0.48</td>\n",
       "      <td>2.25</td>\n",
       "      <td>3.33</td>\n",
       "      <td>7.42</td>\n",
       "      <td>85.42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 804 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CGGA_ID   AASS ABCA13  ABCB1 ABCB4 ABCB5  ABCB8  ABCF2 ABHD11   ACHE  \\\n",
       "320  CGGA_243   7.39   0.02   8.22  1.55   0.0   7.81  24.37  10.29   3.39   \n",
       "321  CGGA_247  18.09   0.29   6.13  1.03  0.09  13.36  22.01  11.25   2.66   \n",
       "322  CGGA_738  22.54   1.01  15.89  1.53  0.19  27.94  28.72  17.74  18.81   \n",
       "323  CGGA_759  16.61   1.08  14.85  1.14   0.0  20.81  21.38   14.9  13.82   \n",
       "324  CGGA_D30   5.57   0.15  14.73  1.89   0.0  14.32  21.45  18.22  12.18   \n",
       "\n",
       "     ... ZNF862 ZNF890P  ZNF92  ZNHIT1 ZNRF2 ZNRF2P1 ZNRF2P2   ZP3 ZSCAN21  \\\n",
       "320  ...   7.75    0.11  14.08   24.01  3.48    0.32    2.47   0.0    7.53   \n",
       "321  ...  48.01    0.85  14.51   29.08  3.64     0.0    0.91  0.74    9.18   \n",
       "322  ...  15.13    0.32   6.47   55.57  5.67    0.69    3.64  0.45   10.07   \n",
       "323  ...  12.43    0.25  11.28   84.91  7.98     0.0    3.13  0.43   10.12   \n",
       "324  ...   7.77    0.27   5.89  114.33  8.57    0.48    2.25  3.33    7.42   \n",
       "\n",
       "       ZYX  \n",
       "320  21.45  \n",
       "321  67.38  \n",
       "322  72.42  \n",
       "323  86.75  \n",
       "324  85.42  \n",
       "\n",
       "[5 rows x 804 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the dataset\n",
    "df_genes = pd.read_csv(\"C:/Users/heung/OneDrive/Documents/scData/cgga_mRNAseq_325/CGGA.mRNAseq_325.RSEM-genes.20200506.txt\", sep=\"\\t\")\n",
    "\n",
    "# Transpose the dataframe\n",
    "df_genes = df_genes.T\n",
    "df_genes.rename(columns = df_genes.iloc[0], inplace = True)\n",
    "df_genes = df_genes.drop(df_genes.index[0])\n",
    "\n",
    "df_genes.reset_index(inplace=True)\n",
    "df_genes.rename(columns = {'index': 'CGGA_ID'}, inplace = True)\n",
    "\n",
    "# Drop the columns that are not needed\n",
    "# df1 = df.drop(columns = ['samples', 'cancer type abbreviation'], inplace = False)\n",
    "# df1.head(5)\n",
    "\n",
    "# Keep the columns that are needed\n",
    "columns_to_keep = [col for col in gain_genes if col in df_genes.columns]\n",
    "\n",
    "df_genes = df_genes[['CGGA_ID'] + columns_to_keep]\n",
    "\n",
    "\n",
    "df_genes.tail(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGGA_ID</th>\n",
       "      <th>PRS_type</th>\n",
       "      <th>Grade</th>\n",
       "      <th>OS</th>\n",
       "      <th>Censor</th>\n",
       "      <th>AASS</th>\n",
       "      <th>ABCA13</th>\n",
       "      <th>ABCB1</th>\n",
       "      <th>ABCB4</th>\n",
       "      <th>ABCB5</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF862</th>\n",
       "      <th>ZNF890P</th>\n",
       "      <th>ZNF92</th>\n",
       "      <th>ZNHIT1</th>\n",
       "      <th>ZNRF2</th>\n",
       "      <th>ZNRF2P1</th>\n",
       "      <th>ZNRF2P2</th>\n",
       "      <th>ZP3</th>\n",
       "      <th>ZSCAN21</th>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>CGGA_1246</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO III</td>\n",
       "      <td>2972.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.16</td>\n",
       "      <td>0.06</td>\n",
       "      <td>11.71</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.02</td>\n",
       "      <td>...</td>\n",
       "      <td>9.32</td>\n",
       "      <td>0.31</td>\n",
       "      <td>12.31</td>\n",
       "      <td>41.66</td>\n",
       "      <td>2.67</td>\n",
       "      <td>0.74</td>\n",
       "      <td>1.11</td>\n",
       "      <td>3.19</td>\n",
       "      <td>10.21</td>\n",
       "      <td>26.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>CGGA_1275</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>183.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.71</td>\n",
       "      <td>0.36</td>\n",
       "      <td>5.61</td>\n",
       "      <td>8.54</td>\n",
       "      <td>0.49</td>\n",
       "      <td>...</td>\n",
       "      <td>7.72</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.47</td>\n",
       "      <td>106.53</td>\n",
       "      <td>4.8</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.33</td>\n",
       "      <td>1.15</td>\n",
       "      <td>6.9</td>\n",
       "      <td>301.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>CGGA_1450</td>\n",
       "      <td>Secondary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>957.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.13</td>\n",
       "      <td>0.02</td>\n",
       "      <td>6.93</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.2</td>\n",
       "      <td>79.41</td>\n",
       "      <td>2.05</td>\n",
       "      <td>1.1</td>\n",
       "      <td>4.52</td>\n",
       "      <td>0.24</td>\n",
       "      <td>8.58</td>\n",
       "      <td>60.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>CGGA_1460</td>\n",
       "      <td>Secondary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>209.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22.76</td>\n",
       "      <td>0.17</td>\n",
       "      <td>6.84</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.77</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4.65</td>\n",
       "      <td>89.46</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2.7</td>\n",
       "      <td>0.37</td>\n",
       "      <td>12.84</td>\n",
       "      <td>243.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>CGGA_1475</td>\n",
       "      <td>Secondary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>182.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.89</td>\n",
       "      <td>1.34</td>\n",
       "      <td>2.96</td>\n",
       "      <td>1.49</td>\n",
       "      <td>0.03</td>\n",
       "      <td>...</td>\n",
       "      <td>10.23</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.89</td>\n",
       "      <td>69.36</td>\n",
       "      <td>5.17</td>\n",
       "      <td>5.92</td>\n",
       "      <td>6.2</td>\n",
       "      <td>0.06</td>\n",
       "      <td>12.03</td>\n",
       "      <td>112.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>CGGA_243</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO II</td>\n",
       "      <td>2977.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.39</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8.22</td>\n",
       "      <td>1.55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.75</td>\n",
       "      <td>0.11</td>\n",
       "      <td>14.08</td>\n",
       "      <td>24.01</td>\n",
       "      <td>3.48</td>\n",
       "      <td>0.32</td>\n",
       "      <td>2.47</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.53</td>\n",
       "      <td>21.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>CGGA_247</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO III</td>\n",
       "      <td>609.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18.09</td>\n",
       "      <td>0.29</td>\n",
       "      <td>6.13</td>\n",
       "      <td>1.03</td>\n",
       "      <td>0.09</td>\n",
       "      <td>...</td>\n",
       "      <td>48.01</td>\n",
       "      <td>0.85</td>\n",
       "      <td>14.51</td>\n",
       "      <td>29.08</td>\n",
       "      <td>3.64</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.74</td>\n",
       "      <td>9.18</td>\n",
       "      <td>67.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>CGGA_738</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO II</td>\n",
       "      <td>4132.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.54</td>\n",
       "      <td>1.01</td>\n",
       "      <td>15.89</td>\n",
       "      <td>1.53</td>\n",
       "      <td>0.19</td>\n",
       "      <td>...</td>\n",
       "      <td>15.13</td>\n",
       "      <td>0.32</td>\n",
       "      <td>6.47</td>\n",
       "      <td>55.57</td>\n",
       "      <td>5.67</td>\n",
       "      <td>0.69</td>\n",
       "      <td>3.64</td>\n",
       "      <td>0.45</td>\n",
       "      <td>10.07</td>\n",
       "      <td>72.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>CGGA_759</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>1263.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16.61</td>\n",
       "      <td>1.08</td>\n",
       "      <td>14.85</td>\n",
       "      <td>1.14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>12.43</td>\n",
       "      <td>0.25</td>\n",
       "      <td>11.28</td>\n",
       "      <td>84.91</td>\n",
       "      <td>7.98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.13</td>\n",
       "      <td>0.43</td>\n",
       "      <td>10.12</td>\n",
       "      <td>86.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>CGGA_D30</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>215.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.57</td>\n",
       "      <td>0.15</td>\n",
       "      <td>14.73</td>\n",
       "      <td>1.89</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.77</td>\n",
       "      <td>0.27</td>\n",
       "      <td>5.89</td>\n",
       "      <td>114.33</td>\n",
       "      <td>8.57</td>\n",
       "      <td>0.48</td>\n",
       "      <td>2.25</td>\n",
       "      <td>3.33</td>\n",
       "      <td>7.42</td>\n",
       "      <td>85.42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 808 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CGGA_ID   PRS_type    Grade      OS  Censor   AASS ABCA13  ABCB1 ABCB4  \\\n",
       "315  CGGA_1246    Primary  WHO III  2972.0     0.0  12.16   0.06  11.71  0.34   \n",
       "316  CGGA_1275    Primary   WHO IV   183.0     1.0   8.71   0.36   5.61  8.54   \n",
       "317  CGGA_1450  Secondary   WHO IV   957.0     1.0   7.13   0.02   6.93  0.77   \n",
       "318  CGGA_1460  Secondary   WHO IV   209.0     1.0  22.76   0.17   6.84  0.95   \n",
       "319  CGGA_1475  Secondary   WHO IV   182.0     1.0  12.89   1.34   2.96  1.49   \n",
       "320   CGGA_243    Primary   WHO II  2977.0     0.0   7.39   0.02   8.22  1.55   \n",
       "321   CGGA_247    Primary  WHO III   609.0     1.0  18.09   0.29   6.13  1.03   \n",
       "322   CGGA_738    Primary   WHO II  4132.0     0.0  22.54   1.01  15.89  1.53   \n",
       "323   CGGA_759    Primary   WHO IV  1263.0     1.0  16.61   1.08  14.85  1.14   \n",
       "324   CGGA_D30    Primary   WHO IV   215.0     1.0   5.57   0.15  14.73  1.89   \n",
       "\n",
       "    ABCB5  ... ZNF862 ZNF890P  ZNF92  ZNHIT1 ZNRF2 ZNRF2P1 ZNRF2P2   ZP3  \\\n",
       "315  0.02  ...   9.32    0.31  12.31   41.66  2.67    0.74    1.11  3.19   \n",
       "316  0.49  ...   7.72     0.0   1.47  106.53   4.8     0.6    0.33  1.15   \n",
       "317   0.0  ...   11.5     0.0    4.2   79.41  2.05     1.1    4.52  0.24   \n",
       "318   0.0  ...  13.77     0.5   4.65   89.46   2.8    0.52     2.7  0.37   \n",
       "319  0.03  ...  10.23     0.0   8.89   69.36  5.17    5.92     6.2  0.06   \n",
       "320   0.0  ...   7.75    0.11  14.08   24.01  3.48    0.32    2.47   0.0   \n",
       "321  0.09  ...  48.01    0.85  14.51   29.08  3.64     0.0    0.91  0.74   \n",
       "322  0.19  ...  15.13    0.32   6.47   55.57  5.67    0.69    3.64  0.45   \n",
       "323   0.0  ...  12.43    0.25  11.28   84.91  7.98     0.0    3.13  0.43   \n",
       "324   0.0  ...   7.77    0.27   5.89  114.33  8.57    0.48    2.25  3.33   \n",
       "\n",
       "    ZSCAN21     ZYX  \n",
       "315   10.21   26.78  \n",
       "316     6.9  301.38  \n",
       "317    8.58   60.35  \n",
       "318   12.84   243.2  \n",
       "319   12.03  112.77  \n",
       "320    7.53   21.45  \n",
       "321    9.18   67.38  \n",
       "322   10.07   72.42  \n",
       "323   10.12   86.75  \n",
       "324    7.42   85.42  \n",
       "\n",
       "[10 rows x 808 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df = pd.merge(df_clinical, df_genes, on='CGGA_ID', how = 'inner')\n",
    "merged_df = merged_df.dropna() # remove NaN\n",
    "merged_df.tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGGA_ID</th>\n",
       "      <th>PRS_type</th>\n",
       "      <th>Grade</th>\n",
       "      <th>OS</th>\n",
       "      <th>Censor</th>\n",
       "      <th>AASS</th>\n",
       "      <th>ABCA13</th>\n",
       "      <th>ABCB1</th>\n",
       "      <th>ABCB4</th>\n",
       "      <th>ABCB5</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF862</th>\n",
       "      <th>ZNF890P</th>\n",
       "      <th>ZNF92</th>\n",
       "      <th>ZNHIT1</th>\n",
       "      <th>ZNRF2</th>\n",
       "      <th>ZNRF2P1</th>\n",
       "      <th>ZNRF2P2</th>\n",
       "      <th>ZP3</th>\n",
       "      <th>ZSCAN21</th>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGGA_1001</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>3817.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.89</td>\n",
       "      <td>0.68</td>\n",
       "      <td>1.82</td>\n",
       "      <td>1.29</td>\n",
       "      <td>0.08</td>\n",
       "      <td>...</td>\n",
       "      <td>9.81</td>\n",
       "      <td>0.23</td>\n",
       "      <td>4.72</td>\n",
       "      <td>50.52</td>\n",
       "      <td>13.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.82</td>\n",
       "      <td>1.75</td>\n",
       "      <td>8.76</td>\n",
       "      <td>158.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGGA_1006</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO III</td>\n",
       "      <td>254.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.84</td>\n",
       "      <td>1.27</td>\n",
       "      <td>14.0</td>\n",
       "      <td>4.52</td>\n",
       "      <td>0.92</td>\n",
       "      <td>...</td>\n",
       "      <td>15.06</td>\n",
       "      <td>0.33</td>\n",
       "      <td>7.73</td>\n",
       "      <td>150.96</td>\n",
       "      <td>4.14</td>\n",
       "      <td>0.94</td>\n",
       "      <td>2.54</td>\n",
       "      <td>9.17</td>\n",
       "      <td>11.94</td>\n",
       "      <td>761.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CGGA_1007</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>345.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.57</td>\n",
       "      <td>0.19</td>\n",
       "      <td>28.82</td>\n",
       "      <td>9.35</td>\n",
       "      <td>0.01</td>\n",
       "      <td>...</td>\n",
       "      <td>16.46</td>\n",
       "      <td>0.22</td>\n",
       "      <td>5.23</td>\n",
       "      <td>59.32</td>\n",
       "      <td>5.58</td>\n",
       "      <td>3.01</td>\n",
       "      <td>5.95</td>\n",
       "      <td>2.04</td>\n",
       "      <td>8.68</td>\n",
       "      <td>92.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGGA_1011</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>109.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.4</td>\n",
       "      <td>0.78</td>\n",
       "      <td>16.15</td>\n",
       "      <td>6.25</td>\n",
       "      <td>0.23</td>\n",
       "      <td>...</td>\n",
       "      <td>8.19</td>\n",
       "      <td>0.45</td>\n",
       "      <td>4.21</td>\n",
       "      <td>39.79</td>\n",
       "      <td>10.62</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.06</td>\n",
       "      <td>0.66</td>\n",
       "      <td>6.64</td>\n",
       "      <td>158.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGGA_1015</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>164.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.56</td>\n",
       "      <td>1.97</td>\n",
       "      <td>1.83</td>\n",
       "      <td>1.51</td>\n",
       "      <td>0.33</td>\n",
       "      <td>...</td>\n",
       "      <td>11.24</td>\n",
       "      <td>0.33</td>\n",
       "      <td>7.31</td>\n",
       "      <td>72.69</td>\n",
       "      <td>7.89</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.61</td>\n",
       "      <td>1.94</td>\n",
       "      <td>4.94</td>\n",
       "      <td>380.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>CGGA_243</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO II</td>\n",
       "      <td>2977.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.39</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8.22</td>\n",
       "      <td>1.55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.75</td>\n",
       "      <td>0.11</td>\n",
       "      <td>14.08</td>\n",
       "      <td>24.01</td>\n",
       "      <td>3.48</td>\n",
       "      <td>0.32</td>\n",
       "      <td>2.47</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.53</td>\n",
       "      <td>21.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>CGGA_247</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO III</td>\n",
       "      <td>609.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18.09</td>\n",
       "      <td>0.29</td>\n",
       "      <td>6.13</td>\n",
       "      <td>1.03</td>\n",
       "      <td>0.09</td>\n",
       "      <td>...</td>\n",
       "      <td>48.01</td>\n",
       "      <td>0.85</td>\n",
       "      <td>14.51</td>\n",
       "      <td>29.08</td>\n",
       "      <td>3.64</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.74</td>\n",
       "      <td>9.18</td>\n",
       "      <td>67.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>CGGA_738</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO II</td>\n",
       "      <td>4132.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.54</td>\n",
       "      <td>1.01</td>\n",
       "      <td>15.89</td>\n",
       "      <td>1.53</td>\n",
       "      <td>0.19</td>\n",
       "      <td>...</td>\n",
       "      <td>15.13</td>\n",
       "      <td>0.32</td>\n",
       "      <td>6.47</td>\n",
       "      <td>55.57</td>\n",
       "      <td>5.67</td>\n",
       "      <td>0.69</td>\n",
       "      <td>3.64</td>\n",
       "      <td>0.45</td>\n",
       "      <td>10.07</td>\n",
       "      <td>72.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>CGGA_759</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>1263.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16.61</td>\n",
       "      <td>1.08</td>\n",
       "      <td>14.85</td>\n",
       "      <td>1.14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>12.43</td>\n",
       "      <td>0.25</td>\n",
       "      <td>11.28</td>\n",
       "      <td>84.91</td>\n",
       "      <td>7.98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.13</td>\n",
       "      <td>0.43</td>\n",
       "      <td>10.12</td>\n",
       "      <td>86.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>CGGA_D30</td>\n",
       "      <td>Primary</td>\n",
       "      <td>WHO IV</td>\n",
       "      <td>215.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.57</td>\n",
       "      <td>0.15</td>\n",
       "      <td>14.73</td>\n",
       "      <td>1.89</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.77</td>\n",
       "      <td>0.27</td>\n",
       "      <td>5.89</td>\n",
       "      <td>114.33</td>\n",
       "      <td>8.57</td>\n",
       "      <td>0.48</td>\n",
       "      <td>2.25</td>\n",
       "      <td>3.33</td>\n",
       "      <td>7.42</td>\n",
       "      <td>85.42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>309 rows × 808 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CGGA_ID PRS_type    Grade      OS  Censor   AASS ABCA13  ABCB1 ABCB4  \\\n",
       "0    CGGA_1001  Primary   WHO IV  3817.0     0.0   6.89   0.68   1.82  1.29   \n",
       "1    CGGA_1006  Primary  WHO III   254.0     1.0   5.84   1.27   14.0  4.52   \n",
       "2    CGGA_1007  Primary   WHO IV   345.0     1.0  10.57   0.19  28.82  9.35   \n",
       "3    CGGA_1011  Primary   WHO IV   109.0     1.0   11.4   0.78  16.15  6.25   \n",
       "4    CGGA_1015  Primary   WHO IV   164.0     1.0   8.56   1.97   1.83  1.51   \n",
       "..         ...      ...      ...     ...     ...    ...    ...    ...   ...   \n",
       "320   CGGA_243  Primary   WHO II  2977.0     0.0   7.39   0.02   8.22  1.55   \n",
       "321   CGGA_247  Primary  WHO III   609.0     1.0  18.09   0.29   6.13  1.03   \n",
       "322   CGGA_738  Primary   WHO II  4132.0     0.0  22.54   1.01  15.89  1.53   \n",
       "323   CGGA_759  Primary   WHO IV  1263.0     1.0  16.61   1.08  14.85  1.14   \n",
       "324   CGGA_D30  Primary   WHO IV   215.0     1.0   5.57   0.15  14.73  1.89   \n",
       "\n",
       "    ABCB5  ... ZNF862 ZNF890P  ZNF92  ZNHIT1  ZNRF2 ZNRF2P1 ZNRF2P2   ZP3  \\\n",
       "0    0.08  ...   9.81    0.23   4.72   50.52   13.2     0.0    1.82  1.75   \n",
       "1    0.92  ...  15.06    0.33   7.73  150.96   4.14    0.94    2.54  9.17   \n",
       "2    0.01  ...  16.46    0.22   5.23   59.32   5.58    3.01    5.95  2.04   \n",
       "3    0.23  ...   8.19    0.45   4.21   39.79  10.62     0.0    1.06  0.66   \n",
       "4    0.33  ...  11.24    0.33   7.31   72.69   7.89    0.26    0.61  1.94   \n",
       "..    ...  ...    ...     ...    ...     ...    ...     ...     ...   ...   \n",
       "320   0.0  ...   7.75    0.11  14.08   24.01   3.48    0.32    2.47   0.0   \n",
       "321  0.09  ...  48.01    0.85  14.51   29.08   3.64     0.0    0.91  0.74   \n",
       "322  0.19  ...  15.13    0.32   6.47   55.57   5.67    0.69    3.64  0.45   \n",
       "323   0.0  ...  12.43    0.25  11.28   84.91   7.98     0.0    3.13  0.43   \n",
       "324   0.0  ...   7.77    0.27   5.89  114.33   8.57    0.48    2.25  3.33   \n",
       "\n",
       "    ZSCAN21     ZYX  \n",
       "0      8.76  158.36  \n",
       "1     11.94  761.56  \n",
       "2      8.68   92.85  \n",
       "3      6.64  158.75  \n",
       "4      4.94  380.73  \n",
       "..      ...     ...  \n",
       "320    7.53   21.45  \n",
       "321    9.18   67.38  \n",
       "322   10.07   72.42  \n",
       "323   10.12   86.75  \n",
       "324    7.42   85.42  \n",
       "\n",
       "[309 rows x 808 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = pd.DataFrame(index = columns_to_keep)\n",
    "result_df['log_rank'] = np.nan\n",
    "result_df['p_value'] = np.nan\n",
    "result_df['survival'] = np.nan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LOG RANK analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "from lifelines.statistics import logrank_test\n",
    "\n",
    "def survival_analysis(gene):\n",
    "    # split the group\n",
    "    quartiles = pd.qcut(merged_df[gene], q = 2, labels = ['low', 'high'])\n",
    "    low_df = merged_df[quartiles == 'low']\n",
    "    high_df = merged_df[quartiles == 'high']\n",
    "    \n",
    "    result = logrank_test(low_df['OS'],\n",
    "                        high_df['OS'],\n",
    "                        event_observed_A = low_df['Censor'],\n",
    "                        event_observed_B = high_df['Censor'])\n",
    "\n",
    "    median_high = high_df['OS'].median()\n",
    "    median_low = low_df['OS'].median()\n",
    "\n",
    "    result_df.at[gene, 'log_rank'] = result.test_statistic\n",
    "    result_df.at[gene, 'p_value'] = result.p_value\n",
    "    \n",
    "    if median_high > median_low:\n",
    "        result_df.at[gene, 'survival'] = 'high'\n",
    "    elif median_high < median_low:\n",
    "        result_df.at[gene, 'survival'] = 'low'\n",
    "    else:\n",
    "        result_df.at[gene, 'survival'] = 'same'\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in columns_to_keep:\n",
    "    survival_analysis(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log_rank</th>\n",
       "      <th>p_value</th>\n",
       "      <th>survival</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AASS</th>\n",
       "      <td>2.269337</td>\n",
       "      <td>1.319564e-01</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA13</th>\n",
       "      <td>29.136823</td>\n",
       "      <td>6.744289e-08</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB1</th>\n",
       "      <td>12.291567</td>\n",
       "      <td>4.550095e-04</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>41.539296</td>\n",
       "      <td>1.155253e-10</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB5</th>\n",
       "      <td>7.705491</td>\n",
       "      <td>5.505310e-03</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF2P1</th>\n",
       "      <td>2.168065</td>\n",
       "      <td>1.409035e-01</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF2P2</th>\n",
       "      <td>7.891946</td>\n",
       "      <td>4.965540e-03</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZP3</th>\n",
       "      <td>12.113750</td>\n",
       "      <td>5.005137e-04</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN21</th>\n",
       "      <td>10.730151</td>\n",
       "      <td>1.054038e-03</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>103.947021</td>\n",
       "      <td>2.077988e-24</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>803 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           log_rank       p_value survival\n",
       "AASS       2.269337  1.319564e-01      low\n",
       "ABCA13    29.136823  6.744289e-08      low\n",
       "ABCB1     12.291567  4.550095e-04     high\n",
       "ABCB4     41.539296  1.155253e-10      low\n",
       "ABCB5      7.705491  5.505310e-03      low\n",
       "...             ...           ...      ...\n",
       "ZNRF2P1    2.168065  1.409035e-01     high\n",
       "ZNRF2P2    7.891946  4.965540e-03     high\n",
       "ZP3       12.113750  5.005137e-04      low\n",
       "ZSCAN21   10.730151  1.054038e-03      low\n",
       "ZYX      103.947021  2.077988e-24      low\n",
       "\n",
       "[803 rows x 3 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log_rank</th>\n",
       "      <th>p_value</th>\n",
       "      <th>survival</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCA13</th>\n",
       "      <td>29.136823</td>\n",
       "      <td>6.744289e-08</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB1</th>\n",
       "      <td>12.291567</td>\n",
       "      <td>4.550095e-04</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>41.539296</td>\n",
       "      <td>1.155253e-10</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCF2</th>\n",
       "      <td>25.010191</td>\n",
       "      <td>5.702810e-07</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACN9</th>\n",
       "      <td>31.892828</td>\n",
       "      <td>1.629176e-08</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF853</th>\n",
       "      <td>11.268696</td>\n",
       "      <td>7.882507e-04</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNHIT1</th>\n",
       "      <td>49.363542</td>\n",
       "      <td>2.126613e-12</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF2</th>\n",
       "      <td>55.089260</td>\n",
       "      <td>1.151785e-13</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZP3</th>\n",
       "      <td>12.113750</td>\n",
       "      <td>5.005137e-04</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>103.947021</td>\n",
       "      <td>2.077988e-24</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>439 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          log_rank       p_value survival\n",
       "ABCA13   29.136823  6.744289e-08      low\n",
       "ABCB1    12.291567  4.550095e-04     high\n",
       "ABCB4    41.539296  1.155253e-10      low\n",
       "ABCF2    25.010191  5.702810e-07      low\n",
       "ACN9     31.892828  1.629176e-08      low\n",
       "...            ...           ...      ...\n",
       "ZNF853   11.268696  7.882507e-04     high\n",
       "ZNHIT1   49.363542  2.126613e-12      low\n",
       "ZNRF2    55.089260  1.151785e-13      low\n",
       "ZP3      12.113750  5.005137e-04      low\n",
       "ZYX     103.947021  2.077988e-24      low\n",
       "\n",
       "[439 rows x 3 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_result_df = result_df[result_df['p_value'] < 0.001]\n",
    "filtered_result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_result_df.to_csv('gain_geneset_survival_analysis.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jswenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
